platform/mellanox/mlnx-platform-api/sonic_platform/component.py

```diff
@@ -101,15 +128,63 @@ def get_firmware_version(self):
         """
         bios_ver_str = self._get_command_result(BIOS_QUERY_VERSION_COMMAND)
         try:
-            m = re.search(self.BIOS_VERSION_PARSE_PATTERN, bios_ver_str)
+            m = re.search(BIOS_VERSION_PARSE_PATTERN, bios_ver_str)
             result = m.group(1)
         except AttributeError as e:
             raise RuntimeError("Failed to parse BIOS version by {} from {} due to {}".format(
-                               self.BIOS_VERSION_PARSE_PATTERN, bios_ver_str, repr(e)))
+                               BIOS_VERSION_PARSE_PATTERN, bios_ver_str, repr(e)))
 
         return result
 
 
+    def install_firmware(self, image_path):
+        """
+        Installs firmware to the component
+
+        Args:
+            image_path: A string, path to firmware image
+
+        Returns:
+            A boolean, True if install was successful, False if not
+        """
+        # check ONIE version. To update ONIE requires version 5.2.0016 or later.
+        machine_conf = self._read_generic_file(MACHINE_CONF_FILE, MACHINE_CONF_MAX_SIZE)
+        try:
+            m = re.search(ONIE_VERSION_PARSE_PATTERN, machine_conf)
+            onie_major = m.group(ONIE_VERSION_MAJOR_OFFSET)
+            onie_minor = m.group(ONIE_VERSION_MINOR_OFFSET)
+            onie_release = m.group(ONIE_VERSION_RELEASE_OFFSET)
+        except AttributeError as e:
+            print("Failed to parse ONIE version by {} from {} due to {}".format(
+                                BIOS_VERSION_PARSE_PATTERN, machine_conf, repr(e)))
+            return False
+        if onie_major < ONIE_REQUIRED_MAJOR or onie_minor < ONIE_REQUIRED_MINOR or onie_release < ONIE_REQUIRED_RELEASE:
+            print("ONIE {}.{}.{} or later is required".format(ONIE_REQUIRED_MAJOR, ONIE_REQUIRED_MINOR, ONIE_REQUIRED_RELEASE))
+            return False
+
+        # check whether the file exists
+        image_exists, message = self._does_file_exist(image_path)
+        if not image_exists:
+            print(message)
+            return False
+
+        # do the real work.
+        try:
+            result = subprocess.call(ONIE_FW_UPDATE_CMD_ADD.format(image_path).split())
+            if result:
+                return False
+            result = subprocess.call(ONIE_FW_UPDATE_CMD_UPDATE.split())
```

Comment:
@stephenxs What if i call another BIOS firmware install before the reboot? Will it overwrite previous? Will it show some message to user?

Reply:
to replace the old one.